### PR TITLE
Project selector api top bar widget with handling options 

### PR DESCRIPTION
**Pull request by jessieyu777**
- Fetching Data from custom hooks & Rendering Projects Data into TopBarWidget

Screenshot 2024-09-10 at 9 58 18 PM

- Adding function handler in TopBarWidget

Screenshot 2024-09-10 at 9 58 25 PM

-preventing rendering the page til the projects are loaded

### CODE REVIEW

File: apps/gitness/src/pages/pipeline-list.tsx [hunk 1]
```diff
@@ -62,9 +72,26 @@ export default function PipelinesPage() {
     )
   }
 
+  //fetch projects api to get the list of projects
+  const { data: projects, isLoading } = useMembershipSpacesQuery({
+    queryParams: { page: 1, limit: 30, sort: 'identifier', order: 'asc' }
+  })
+
+  //prevent rendering the page until the projects are loaded
+  if (isLoading) {
```
Comment (joetaylor-harness) [inline apps/gitness/src/pages/pipeline-list.tsx:81]: @jessieyu777 blocking the entire page from loading while the project list loads could be a bit too dramatic? Maybe the dropdown is where the project loading state should be communicated?

**Review by joetaylor-harness — COMMENTED** [review]
@jessieyu777 the playground changes for TopBarWidgets LGTM, I just have the question about loading state blocking the page render in gitness.

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 1 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx]: @joetaylor-harness That's a temporary status for not breaking the page from loading because of the gap of getting data. The error will show from the component side since if the projects props was empty. so I wrote a condition function to the high level to this page first. Already discussed it in the meeting today. We decided to make this branch merge first, and then figure out what is the best status for showing this loading since we didn't get the design-look for this details.

File: apps/gitness/src/pages/pipeline-list.tsx [hunk 2]
```diff
@@ -13,9 +13,25 @@ import {
...
+
+interface Project {
+  id: string | undefined
+  identifier: string | undefined
+}
+
+interface Project {
+  id: string | undefined
+  identifier: string | undefined
+}
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/pages/pipeline-list.tsx:34]: duplicate interface.. how did typescript and eslint not catch this?

also, why do we have a hand-coded interface here? all API interfaces should be coming from the service client

File: apps/gitness/src/pages/pipeline-list.tsx [hunk 3]
```diff
@@ -37,6 +53,7 @@ export default function PipelinesPage() {
     }
   )
 
+  const LinkComponent = ({ to, children }: { to: string; children: React.ReactNode }) => <Link to={to}>{children}</Link>
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/pages/pipeline-list.tsx:56]: is this your change? seems unrelated

File: apps/gitness/src/pages/pipeline-list.tsx [hunk 4]
```diff
@@ -58,13 +75,35 @@ export default function PipelinesPage() {
...
 
+  //fetch projects api to get the list of projects
+  const { data: projects, isLoading } = useMembershipSpacesQuery({
+    queryParams: { page: 1, limit: 30, sort: 'identifier', order: 'asc' }
+  })
+
+  const projectsItem: Project[] =
+    // @ts-expect-error remove "@ts-expect-error" once type issue for "content" is resolved
+    projects?.content?.map(membership => ({
+      id: `${membership?.space?.id}`,
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/pages/pipeline-list.tsx:91]: unnecessary template literals
```suggestion
      id: membership?.space?.id,
```

File: apps/gitness/src/pages/pipeline-list.tsx [hunk 5]
```diff
@@ -58,13 +75,35 @@ export default function PipelinesPage() {
...
+      //not sure the design state of the loading, this is temoporary
+      return <div className="w-full h-full text-center">isLoading...</div>
+    }
+    return <TopBarWidget projects={projectsItem || []} />
+  }
+
   return (
     <>
-      <TopBarWidget />
+      {renderTopBar()}
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/pages/pipeline-list.tsx:106]: I'd prefer to use component syntax instead of function syntax

File: packages/playground/src/assets/pipelineV0.yaml [hunk 6]
```diff
@@ -0,0 +1,121 @@
+pipeline:
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/assets/pipelineV0.yaml:1]: unrelated file?

File: packages/playground/src/components/layout/top-bar-widget.tsx [hunk 7]
```diff
@@ -33,32 +33,44 @@ interface BreadcrumbItemProps {
...
-    </DropdownMenuTrigger>
-    <DropdownMenuContent align="start" className="mt-1.5">
-      {projects.map(project => (
-        <DropdownMenuItem key={project.id}>{project.name}</DropdownMenuItem>
-      ))}
-    </DropdownMenuContent>
-  </DropdownMenu>
-)
+const ProjectDropdown: React.FC<{ isPrimary: boolean; projects: Project[] }> = ({ isPrimary, projects }) => {
+  const [selectedProject, setSelectedProject] = useState<string>(projects[0]?.name || '')
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/components/layout/top-bar-widget.tsx:37]: default should not be the first project.. default should be "no project selected"

File: packages/playground/src/components/layout/top-bar-widget.tsx [hunk 8]
```diff
@@ -33,32 +33,44 @@ interface BreadcrumbItemProps {
...
+          size={10}
+          className={cn('chevron-down', {
+            'text-primary': isPrimary,
+            'text-navbar-text-secondary group-hover:text-primary': !isPrimary
+          })}
+        />
+      </DropdownMenuTrigger>
+      <DropdownMenuContent align="start" className="mt-1.5">
+        {projects.length === 0
+          ? 'No Project'
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/components/layout/top-bar-widget.tsx:64]:
```suggestion
          ? 'No Projects found'
```

File: packages/playground/src/components/pull-request/utils.tsx [hunk 9]
```diff
@@ -60,46 +60,6 @@ export function useDateFilters() {
     []
   )
 }
-// The timeDistance function calculates the time difference between two dates (in milliseconds)
```
Comment (abhinavrastogi-harness) [inline packages/playground/src/components/pull-request/utils.tsx:63]: unrelated?

**Review by abhinavrastogi-harness — CHANGES_REQUESTED** [review]

File: packages/playground/src/components/pull-request/utils.tsx (hunk 9 above)
Comment (jessieyu777) [reply packages/playground/src/components/pull-request/utils.tsx:63]: Yes unrelated

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 3 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx:56]: no this is not mine

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 2 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx:34]: Got it! I will check on that if they have related interface there

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 4 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx:91]: Since the api is return the number type of id but it is required by a string here. I will change the type in the playground.

File: packages/playground/src/assets/pipelineV0.yaml (hunk 6 above)
Comment (jessieyu777) [reply packages/playground/src/assets/pipelineV0.yaml:1]: Y

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 1 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx]: Currently it look like this since I call api separately. 
Screenshot 2024-09-11 at 2 15 39 PM

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 4 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx]: has been solved the type into the component in the playground

File: apps/gitness/src/pages/pipeline-list.tsx (hunk 5 above)
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx]: move all rendering in components/header-api

File: packages/playground/src/components/pull-request/utils.tsx (hunk 9 above)
Comment (jessieyu777) [reply packages/playground/src/components/pull-request/utils.tsx]: removed

File: packages/playground/src/components/layout/top-bar-widget.tsx [hunk 10]
```diff
@@ -19,7 +19,7 @@ import {
 import { getInitials } from '../../utils/utils'
 
 interface Project {
-  id: string
+  id: number
   name: string
 }
```
Comment (vardanbansal-harness) [inline packages/playground/src/components/layout/top-bar-widget.tsx:24]: Don't add this new interface, instead use interface [TypesSpace](https://github.com/harness/react-api-client/blob/main/packages/code-service/src/services/schemas/TypesSpace.ts) from `@harnessio/code-service-client`

`import { TypesSpace } from '@harnessio/code-service-client'`

File: packages/playground/src/components/layout/top-bar-widget.tsx [hunk 11]
```diff
@@ -33,32 +33,44 @@ interface BreadcrumbItemProps {
...
+          size={10}
+          className={cn('chevron-down', {
+            'text-primary': isPrimary,
+            'text-navbar-text-secondary group-hover:text-primary': !isPrimary
+          })}
+        />
+      </DropdownMenuTrigger>
+      <DropdownMenuContent align="start" className="mt-1.5">
+        {projects.length === 0
+          ? 'No Project Found'
```
Comment (vardanbansal-harness) [inline packages/playground/src/components/layout/top-bar-widget.tsx:64]: nit: `No Projects Found`

File: apps/gitness/src/components/header-api.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
+import { Topbar, Text } from '@harnessio/canary'
+import { TopBarWidget } from '@harnessio/playground'
+import { useMembershipSpacesQuery } from '@harnessio/code-service-client'
+
+export default function HeaderApi() {
```
Comment (vardanbansal-harness) [inline apps/gitness/src/components/header-api.tsx:5]: nit: Just `Header` would be better.

File: apps/gitness/src/components/header-api.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  const { data: projects, isLoading } = useMembershipSpacesQuery({
+    queryParams: { page: 1, limit: 30, sort: 'identifier', order: 'asc' }
+  })
+  //prevent rendering the page until the projects are loaded
+  if (isLoading) {
+    return (
+      <Topbar.Root>
+        <Topbar.Left>
+          <Text size={2} weight="medium" className="text-primary">
+            Loading Your Project...
```
Comment (vardanbansal-harness) [inline apps/gitness/src/components/header-api.tsx:16]: User can have multiple projects, `Loading your projects...` would be better.

File: apps/gitness/src/components/header-api.tsx (hunk 12 above)
Comment (jessieyu777) [reply apps/gitness/src/components/header-api.tsx]: Solved

File: apps/gitness/src/pages/pipeline-list.tsx [hunk 14]
```diff
@@ -58,13 +75,35 @@ export default function PipelinesPage() {
             }
           ]
         }))}
+        LinkComponent={LinkComponent}
       />
     )
   }
 
+  //fetch projects api to get the list of projects
+  const { data: projects, isLoading } = useMembershipSpacesQuery({
```
[the remark this answers is not in the document]
Comment (jessieyu777) [reply apps/gitness/src/pages/pipeline-list.tsx]: remove it and move to components/header components

File: packages/playground/src/components/layout/top-bar-widget.tsx (hunk 11 above)
Comment (jessieyu777) [reply packages/playground/src/components/layout/top-bar-widget.tsx]: Solve in new commit

**Review by vardanbansal-harness — APPROVED** [review]

**Review by jessieyu777 — COMMENTED** [review]
- Already mock loading status to the feedback of design team. This is temporary status below.

Screenshot 2024-09-11 at 2 15 39 PM
current dropdown status
Screenshot 2024-09-11 at 2 59 55 PM
-changing the word in please select a project instead
Screenshot 2024-09-11 at 4 29 20 PM

**Review by abhinavrastogi-harness — APPROVED** [review]
